### PR TITLE
feat(behavior_path_planner): add function to create PredictedPath from target velocity

### PR DESCRIPTION
**Pull request by kyoichi-sugahara**
## Description

Create a predicted path from the current velocity to a target velocity.
It assume constant acceralation and stopping time is considered if the vehicle is stopped.

image

## Tests performed

check with simple planning simulator

Screenshot from 2023-06-13 14-19-02

## Effects on system behavior

Not applicable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/utils/path_utils.cpp [hunk 1]
```diff
@@ -142,6 +142,67 @@ PathWithLaneId resamplePathWithSpline(
...
+{
+  // Check if following_trajectory_points is empty
+  if (following_trajectory_points.empty()) {
+    return PredictedPath();
+  }
+
+  size_t nearest_segment_index =
+    motion_utils::findNearestIndex(following_trajectory_points, pose.position);
+  FrenetPoint vehicle_pose_frenet =
+    convertToFrenetPoint(following_trajectory_points, pose.position, nearest_segment_index);
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/utils/path_utils.cpp:159]:
```suggestion
  const size_t nearest_segment_index =
    motion_utils::findNearestIndex(following_trajectory_points, pose.position);
  const FrenetPoint vehicle_pose_frenet =
    convertToFrenetPoint(following_trajectory_points, pose.position, nearest_segment_index);
```

File: planning/behavior_path_planner/src/utils/path_utils.cpp [hunk 2]
```diff
@@ -142,6 +142,67 @@ PathWithLaneId resamplePathWithSpline(
...
+      const double delta_t = t - start_time;
+      const double length = initial_velocity * delta_t + 0.5 * acc * delta_t * delta_t + offset;
+      const auto pose = motion_utils::calcInterpolatedPose(
+        following_trajectory_points, vehicle_pose_frenet.length + length);
+      predicted_path.path.push_back(pose);
+    }
+  };
+
+  // Calculate time required to reach target velocity
+  double acc_time = std::max(0.0, (target_velocity - current_velocity) / acc_till_target_velocity);
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/utils/path_utils.cpp:182]:
```suggestion
  const double acc_time = std::max(0.0, (target_velocity - current_velocity) / acc_till_target_velocity);
```

File: planning/behavior_path_planner/src/utils/path_utils.cpp [hunk 3]
```diff
@@ -142,6 +142,67 @@ PathWithLaneId resamplePathWithSpline(
...
+  predicted_path.time_step = rclcpp::Duration::from_seconds(resolution);
+  predicted_path.path.reserve(
+    std::min(following_trajectory_points.size(), static_cast<size_t>(100)));
+
+  // add a segment to the path
+  auto addSegment = [&](
+                      const std::vector<PathPointWithLaneId> & following_trajectory_points,
+                      PredictedPath & predicted_path, const double initial_velocity,
+                      const double acc, const double start_time, const double end_time,
+                      double & offset) {
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/utils/path_utils.cpp:171]:
```suggestion
  auto addSegment = [&](
                      const std::vector<PathPointWithLaneId> & following_trajectory_points,
                      PredictedPath & predicted_path, const double initial_velocity,
                      const double acc, const double start_time, const double end_time,
                      const double offset) {
```

Comment (kosuke55) [reply planning/behavior_path_planner/src/utils/path_utils.cpp:171]: (and I prefer putting output first)

```
auto addSegment = [&](PredictedPath & predicted_path, .....
```

File: planning/behavior_path_planner/src/utils/path_utils.cpp [hunk 4]
```diff
@@ -142,6 +142,68 @@ PathWithLaneId resamplePathWithSpline(
...
+    std::max(0.0, (target_velocity - current_velocity) / acc_till_target_velocity);
+
+  // If stopping time is greater than zero and current velocity is almost zero, add stopping segment
+  if (current_velocity < 0.01 && stopping_time > 0.0) {
+    double offset = 0.0;  // Initialize offset here as it's not used before this point
+    addSegment(following_trajectory_points, predicted_path, 0.0, 0.0, 0.0, stopping_time, offset);
+  }
+
+  // Add acceleration segment
+  double offset = 0.0;
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/utils/path_utils.cpp:192]: offset is always 0.0? this assumes `current_velocity` = 0.0?

Comment (kosuke55) [reply planning/behavior_path_planner/src/utils/path_utils.cpp:192]: oh, I understood

File: planning/behavior_path_planner/src/utils/path_utils.cpp [hunk 5]
```diff
@@ -142,6 +142,68 @@ PathWithLaneId resamplePathWithSpline(
...
+  addSegment(
+    following_trajectory_points, predicted_path, current_velocity, acc_till_target_velocity,
+    stopping_time, stopping_time + acc_time, offset);
+
+  // Add constant velocity segment
+  offset += current_velocity * acc_time + 0.5 * acc_till_target_velocity * acc_time * acc_time;
+  double constant_velocity_time = stopping_time + acc_time + (target_velocity / resolution);
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/utils/path_utils.cpp:199]:
```suggestion
  const double constant_velocity_time = stopping_time + acc_time + (target_velocity / resolution);
```

**Comment by kyoichi-sugahara** [comment]
updated createPredictedPath in the https://github.com/autowarefoundation/autoware.universe/pull/5014

**Pull request closed without merging**